standard/expressions.md

```diff
@@ -4808,7 +4813,13 @@ It is possible for anonymous function delegates to share some captured variables
 > }
 > ```
 >
-> the three delegates capture the same instance of `x` but separate instances of `y`. *end example*
+> the three delegates capture the same instance of `x` but separate instances of `y`, and the output is:
+>
+> ```console
+> 1 1
+> 2 1
+> 3 1
+> ```
```

Comment:
We need the *end example* restored.

Reply:
@jskeet Clearly, as suggested by the opening sentence, this `F` is different to that in the previous example. I see two possible solutions:

1. Change the into text from "For example, if `F` is changed to" to "For example, if `F` is changed, as follows".
2. Clone the whole of the previous example, replacing its `F` with the new version. As that would declare the delegate type `D` as well, the `additionalFiles:["DvoidNoArgs.cs"]` can be removed, and as that is the only place that file is added  (I think), it can be deleted from the addition-files folder.

Yes, the *end example* marker needs to restored too.

Reply:
Hmm. I tend to think that the spec text was basically ideal as it was before - and I've always wanted the example testing to avoid detracting from the spec text. I think option 1 is probably the best for now, but I'll keep thinking about a way to avoid having the loop in the testable example.

Reply:
Option 1 implemented

Reply:
@jskeet Here's an idea that I have working with a small bit of manual intervention. Here's my test code:

````
> <!-- Example: {template:"code-in-class-lib", name:"InstantiationOfLocalVariables6", IgnoredWarnings:["CS8321"], additionalFiles:["Caller.cs"], inferOutput: true} -->
> ```csharp
> static D[] F()
> {
>     D[] result = new D[3];
>     int x = 0;
>     for (int i = 0; i < 3; i++)
>     {
>         int y = 0;
>         result[i] = () => Console.WriteLine($"{++x} {++y}");
>     }
>     return result;
> }
> ```
>
> the three delegates capture the same instance of `x` but separate instances of `y`, and the output is:
>
> ```console
> 1 1
> 2 1
> 3 1
> ```
>
> *end example*
````

When extracted, the Library.cs file contains the following:

```csharp
<<using directives ...>>

class Class1
{
    static D[] F()
{
    D[] result = new D[3];
    int x = 0;
    for (int i = 0; i < 3; i++)
    {
        int y = 0;
        result[i] = () => Console.WriteLine($"{++x} {++y}");
    }
    return result;
}
}
```

**However, what if the extractor added `partial` to `class Class1` for template code-in-class-lib (and code-in-main)?**

Caller.cs contains the following:

```csharp
delegate void D();

partial class Class1
{
   static void Main()
   {
       foreach (D d in F())
       {
           d();
       }
   }
}
```

This allows us to **supplement** the class from the example by declaring the delegate and adding a `Main`. And because both files contribute to the same class, we don't need `public` accessibility in method `F`'s declaration.

This relies on this template emiting `partial`. Now for all other examples that **don't** need supplementing, we'd have just a single Library.cs or Program.cs file containing a partial class, but that's OK, right?

When I compile/execute this, I use

```console
csc Library.cs Caller.cs -t:exe
```

and the resulting executable is Caller.exe.

However, if we have template code-in-main also emit `partial`, and the example contains `Main` and the supplementary file has support stuff, I'd have something like

```console
csc Program.cs Support.cs -t:exe
```

and the resulting executable would be Program.exe. My concern here is that the logic the tester uses to execute the result might be different for these two scenarios.

Reply:
Yes, I think partial classes would be a great way of doing this. There shouldn't be any need to change the execution code.

I don't want the extractor to have to "understand" the code though. I think we can just have a template for "code in a partial class" and then use an additional file for this example that supplies the Main method. (All such examples would just need to use the same partial class name, which is easy enough.) Effectively the new template would be like code-in-class-lib but with a partial class and an executable output type (instead of library).

I'm happy to do this myself if that description is too hard to follow :)

Reply:
@jskeet I'll take a look at implementing this, but probably not for a few days. In the meantime, I'm going back to the V8 spec for nullable references based on recent private feedback from Mads and Bill.

Reply:
@jon I've created the new template code-in-partial-class, so we're good to go. However, on my first attempt I inadvertently added that template to addition-files, but I don't see how to remove that. Can you please delete the 3-rd-tp-last commit? I then put it in the correct place.